src/shogun/features/CombinedDotFeatures.cpp

```diff
@@ -309,32 +304,35 @@ int32_t CCombinedDotFeatures::get_nnz_features_for_vector(int32_t num) const
 	return result;
 }
 
-SGVector<float64_t> CCombinedDotFeatures::get_subfeature_weights()
+SGVector<float64_t> CCombinedDotFeatures::get_subfeature_weights() const
 {
 	int32_t num_weights = get_num_feature_obj();
 	ASSERT(num_weights > 0)
 
-	float64_t* weights=SG_MALLOC(float64_t, num_weights);
+	SGVector<float64_t> weights(num_weights);
+	std::copy(feature_weights.begin(), feature_weights.end(), weights.vector);
 
-	for (index_t f_idx=0; f_idx<num_weights; f_idx++)
-	{
-		CDotFeatures* f = get_feature_obj(f_idx);
-		weights[f_idx] = f->get_combined_feature_weight();
-		SG_UNREF(f);
-	}
-	return SGVector<float64_t>(weights,num_weights);
+	return weights;
 }
 
-void CCombinedDotFeatures::set_subfeature_weights(SGVector<float64_t> weights)
+void CCombinedDotFeatures::set_subfeature_weights(const SGVector<float64_t>& weights)
 {
 	ASSERT(weights.vlen==get_num_feature_obj())
 
-	for (index_t f_idx=0; f_idx<get_num_feature_obj(); f_idx++)
-	{
-		CDotFeatures* f = get_feature_obj(f_idx);
-		f->set_combined_feature_weight(weights[f_idx]);
-		SG_UNREF(f);
-	}
+	std::copy(
+	    weights.vector, weights.vector + weights.vlen, feature_weights.begin());
+}
+
+float64_t CCombinedDotFeatures::get_subfeature_weight(index_t idx) const
+{
+	ASSERT(idx >= 0 && (size_t)idx < feature_weights.size())
+	return feature_weights[idx];
+}
+
+void CCombinedDotFeatures::set_subfeature_weight(index_t idx, float64_t weight)
+{
+	ASSERT(idx >= 0 && (size_t)idx < feature_weights.size())
```

Comment:
if this is user-facing, it should have an error message that moans how the provided differs from the expected